# -*- coding: utf-8 -*-
"""
Created on Mon Sep 12 15:56:02 2022

@author: arman hossain

!python excel.py
"""

import sys

from PyQt5.QtWidgets import (
    QApplication, QDialog, QTableWidget,QWidget, QMainWindow, QMessageBox, QGroupBox, QLabel, QHeaderView
)
from PyQt5.uic import loadUi
import pickle
import os

class Excel(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        loadUi("excel.ui", self)
        self.sortChanged(0)
        self.setWindowTitle("List View")
        self.connect()
        self.initui()
        self.total_data_label.setText(str(len(self.tableWidget.data))+"         ")
        
    def initui(self):
        header = self.tableWidget.horizontalHeader()       
        header.setSectionResizeMode(1, QHeaderView.Stretch)
        self.initCombo()
    
    def initCombo(self):
        if os.path.exists('./data/2002-2022-vendor-year-count.pickle'):
            self.comboBox.addItems(['All']+list(self.read_data("./data/2002-2022-vendor-year-count.pickle").keys()))
        
    def read_data(self,filename):
        with open(filename, 'rb') as file:
            obj = pickle.load(file)
            return obj
        
    def connect(self):
        self.previous_btn.clicked.connect(self.previous_pressed)
        self.next_btn.clicked.connect(self.next_pressed)
        self.plus_btn.clicked.connect(self.current_plus_pressed)
        self.minus_btn.clicked.connect(self.current_minus_pressed)
        # self.current_btn.clicked.connect(self.current_btn_pressed)
        self.tableWidget.horizontalHeader().sortIndicatorChanged.connect(self.sortChanged)
        self.comboBox.activated.connect(self.comboindexchanged)
        
    def comboindexchanged(self,index):
        vendor = self.comboBox.itemText(index)
        self.tableWidget.filter_dataset(vendor)
        self.tableWidget.goto_page(1)
        self.setbuttontext(1)
        self.total_data_label.setText(str(len(self.tableWidget.data))+"         ")
        
    def sortChanged(self,index):
        self.tableWidget.sortChanged(index)
        self.tableWidget.goto_page(1)
        self.setbuttontext(1)
        
    def reload_data(self):
        self.tableWidget.reload_data()
        self.comboBox.setCurrentIndex(0)
        self.tableWidget.goto_page(1)
        self.setbuttontext(1)
        
    def setbuttontext(self,number):
        if number < 1:
            return
        if number == 1:
            self.previous_btn.setEnabled(False)
        else:
            self.previous_btn.setEnabled(True)
        next = number+self.tableWidget.item_per_page
        pre = number-self.tableWidget.item_per_page
        self.current_btn.setText(str(number))
        self.plus_btn.setText(str(next))
        if pre>0:
            self.minus_btn.setEnabled(True)
            self.minus_btn.setText(str(pre))
        else:
            self.minus_btn.setText(str(0))
            self.minus_btn.setEnabled(False)
        
    def current_plus_pressed(self):
        self.tableWidget.goto_page(int(self.plus_btn.text()))
        self.setbuttontext(int(self.plus_btn.text()))
    def current_minus_pressed(self):
        self.tableWidget.goto_page(int(self.minus_btn.text()))
        self.setbuttontext(int(self.minus_btn.text()))
    def next_pressed(self):
        self.tableWidget.goto_page_next()
        self.setbuttontext(int(self.current_btn.text())+1)
    def previous_pressed(self):
        self.tableWidget.goto_page_pre()
        self.setbuttontext(int(self.current_btn.text())-1)
        


if __name__ == "__main__":
    app = QApplication(sys.argv)
    win = Excel()
    win.show()
    sys.exit(app.exec())